src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.Generated.cs

```diff
@@ -12,6 +14,24 @@ partial class LambdaCompiler
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Maintainability", "CA1505:AvoidUnmaintainableCode"), System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Maintainability", "CA1502:AvoidExcessiveComplexity")]
         private void EmitExpression(Expression node, CompilationFlags flags)
         {
+            // When compling deep trees, we run the risk of triggering a terminating StackOverflowException,
+            // so we probe for sufficient stack here. When we run out of stack, we simply continue the work
+            // on another thread. Note that EnsureSufficientExecutionStack is very fast, so the overhead of
+            // this overflow protection is minimal.
```

Comment:
This would be the same approach as taken by Roslyn (although we do not attempt to borrow another stack. since compiler has an easier option of giving an error)

The assumption is that core algorithms of a given app typically operate at nearly empty stack, so we can easily afford 100 levels of recursion before starting to worry about stack.
If the above assumption is not right and we are almost out of stack even before doing this kind of recursion, we are already in trouble and probing is not going to be reliable anyways.